vowpalwabbit/reduction_stack.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include "reductions_fwd.h"
+
+struct vw;
+
+typedef VW::LEARNER::base_learner* (*reduction_setup_fn)(VW::setup_base_fn&, VW::config::options_i&, vw&);
```

Comment:
setup_fn here is a pure virtual class with instantiations that create their own reduction stack.   Can the same thing can not be achieved if setup_fn decides what stack to instantiate?  This achieves the same thing without adding another parameter to the setup function of each reduction.

Reply:
- The concern of adding an extra parameter is interesting, since in another PR after this one you could easily collapse options and all parameters inside setup_base_fn interface. If you care about parameter count, the count would get reduced to 1 and this 'object' starts to represent exactly what it is: a collection of configs and a magic box that gives you the base learner. You get the side benefit of having specific APIs that hide away VW all global state.
- Yes, you could achieve the same thing with a separate and alternate impls of current setup functions. I'm proposing that you can get for free with a single impl of setup function by breaking that coupling between setup functions all over the place and parse_args.h:setup_base fn. You can override this dependency on runtime by following the proposed interface in this PR.

Reply:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/d5bb16e2227148082aadf48daa6be050f139321d/vowpalwabbit/parse_args.cc#L1415

You can imagine alternate builders with other behaviour, like cached_learner that always returns the learner which it was constructed with like so: https://github.com/VowpalWabbit/vowpal_wabbit/blob/d5bb16e2227148082aadf48daa6be050f139321d/vowpalwabbit/cached_learner.h

Reply:
Makes sense.  setup_fn evolving to incorporate a factory for base reductions and other facilities that a reduction needs seems like a good general direction.  This makes it easier for a reduction to live in a different host outside vw.exe as long as the host provides all the same facilities.... like a factory for the next reduction, a random number generator, log facility, state save/resume facility etc.